azure-iot-provisioning-devicessdk/azure/iot/provisioning/devicessdk/security/sk_security_provider.py

```diff
@@ -0,0 +1,64 @@
+# -------------------------------------------------------------------------
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the MIT License. See License.txt in the project root for
+# license information.
+# --------------------------------------------------------------------------
+"""This module contains a provider that is responsible for providing shared access tokens that will eventually establish
+ the authenticity of devices to Device Provisioning Service.
+"""
+from azure.iot.common.sastoken import SasToken
+
+
+class SymmetricKeySecurityProvider(object):
+    """
+    A provider that is responsible for providing shared access tokens that will eventually establish
+    the authenticity of devices to Device Provisioning Service.
+    """
+
+    def __init__(self, registration_id, symmetric_key, id_scope):
+        """
+        Initialize the symmetric key security client.
+        :param registration_id: The registration ID is used to uniquely identify a device in the Device Provisioning Service.
+        The registration ID is alphanumeric, lowercase string and may contain hyphens.
+        :param symmetric_key: The key which will be used to create the shared access signature token to authenticate
+        the device with the Device Provisioning Service. By default, the Device Provisioning Service creates
+        new symmetric keys with a default length of 32 bytes when new enrollments are saved with the Auto-generate keys
+        option enabled. Users can provide their own symmetric keys for enrollments by disabling this option within
+        16 bytes and 64 bytes and in valid Base64 format.
+        :param id_scope: The ID scope is used to uniquely identify the specific provisioning service the device will
+        register through. The ID scope is assigned to a Device Provisioning Service when it is created by the user and
+        is generated by the service and is immutable, guaranteeing uniqueness.
+        """
+        self._registration_id = registration_id
+        self._symmetric_key = symmetric_key
+        self._id_scope = id_scope
+
+    @property
+    def registration_id(self):
+        """
+        :return: The registration ID is used to uniquely identify a device in the Device Provisioning Service.
+        The registration ID is alphanumeric, lowercase string and may contain hyphens.
+        """
+        return self._registration_id
+
+    @property
+    def id_scope(self):
+        """
+        :return: Host running the Device Provisioning Service.
+        """
+        return self._id_scope
```

Comment:
Why not just let these be public attributes? I don't see a reason they couldn't be dynamically modified.

Reply:
There's no reason for them to change during execution. it's associated with a specific device and DPS instance. not sure how you want this to exist in Python, but we don't want the customers to think they can reuse the same client to register multiple devices. I don't think that's a scenario. 

---
In reply to: [265318495](https://github.com/Azure/azure-iot-sdk-python-preview/pull/59#discussion_r265318495) [](ancestors = 265318495)

Reply:
I mean, these attributes only do something when "create_shared_access_signature" is called. That itself just draws on whatever the current value is - there's no inherent need for the values to stay the same, as they aren't being changed internally, or carrying any kind of intermediate value. No other internal or intermediate values are derived from or depend on them either.

If you change the registration id and and id scope, the object isn't invalidated. "create_shared_access_signature" still works as if the object was instantiated with different values. Nothing breaks. Thus, I don't see the need to go private by convention.

The only time the values should be private by convention is when changing them would put the application in an unexpected state IMO. Otherwise, the Python principle of developers being "responsible adults" should apply. Python is not a defensive language. Thus, unless there's a good reason related to the integrity of the application, I don't believe that attributes should be private by convention.

---
In reply to: [265323518](https://github.com/Azure/azure-iot-sdk-python-preview/pull/59#discussion_r265323518) [](ancestors = 265323518,265318495)